gateway-service/src/main/java/com/ca/mfaas/gateway/filters/post/PageRedirectionFilter.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * This program and the accompanying materials are made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Copyright Contributors to the Zowe Project.
+ */
+package com.ca.mfaas.gateway.filters.post;
+
+import com.ca.mfaas.product.gateway.GatewayConfigProperties;
+import com.ca.mfaas.product.routing.RoutedServices;
+import com.ca.mfaas.product.routing.RoutedServicesUser;
+import com.ca.mfaas.product.routing.ServiceType;
+import com.ca.mfaas.product.routing.transform.TransformService;
+import com.ca.mfaas.product.routing.transform.URLTransformationException;
+import com.netflix.util.Pair;
+import com.netflix.zuul.ZuulFilter;
+import com.netflix.zuul.context.RequestContext;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.cloud.client.ServiceInstance;
+import org.springframework.cloud.client.discovery.DiscoveryClient;
+import org.springframework.http.HttpStatus;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.apache.http.HttpHeaders.LOCATION;
+import static org.springframework.cloud.netflix.zuul.filters.support.FilterConstants.*;
+
+/**
+ * This is a post filter for request which returns status code 3XX
+ * <ul>The filter checks Location in response header, and transforms the url to gateway url if the url satisfies 2 conditions:
+ * <li>Hostname and port of the url are registered in Discovery Service</li>
+ * <li>The url can be matched to gateway url</li>
+ * </ul>
+ */
+@Slf4j
+public class PageRedirectionFilter extends ZuulFilter implements RoutedServicesUser {
+
+    private final DiscoveryClient discovery;
+    private final Map<String, RoutedServices> routedServicesMap = new HashMap<>();
+    private final ConcurrentHashMap<String, String> routeTable = new ConcurrentHashMap<>();
+    private final TransformService transformService;
+
+    /**
+     * Constructor
+     *
+     * @param discovery discovery client
+     * @param hostname  hostname of gateway
+     * @param port      port number of gateway
+     * @param scheme    scheme of gateway
+     */
+    public PageRedirectionFilter(DiscoveryClient discovery, String hostname, String port, String scheme) {
```

Comment:
Hi Cindy, 
Thanks again for continued efforts.

Everything is perfect! My suggestions:
By architecture perspective, 
- For the future the hostname, port and scheme values should not be loaded many times by components. 
- Constructor which has many parameters should be decreased

For this, we have GatewayConfigProperties, 
however, I think it would be great if we let to the spring creating this bean for us, not manually.
You can add this class to the ```com.ca.mfaas.gateway.config``` folder.

```java
package com.ca.mfaas.gateway.config;

import com.ca.mfaas.product.gateway.GatewayConfigProperties;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.stereotype.Component;

@Component
public class GatewayConfig {

    @Bean
    public GatewayConfigProperties getGatewayConfigProperties(@Value("${apiml.service.hostname}")String hostname,
                                                              @Value("${apiml.service.port}")String port,
                                                              @Value("${apiml.service.scheme}")String scheme) {
        return GatewayConfigProperties.builder()
            .scheme(scheme)
            .hostname(hostname + ":" + port)
            .build();
    }


}
```

```java
 /**
     * Constructor
     *
     * @param discovery discovery client
     * @param gatewayConfigProperties gateway properties
     */
    public PageRedirectionFilter(DiscoveryClient discovery, GatewayConfigProperties gatewayConfigProperties) {
        this.discovery = discovery;
        transformService = new TransformService(gatewayConfigProperties);
    }
```

And Your tests by above code: ```com.ca.mfaas.gateway.filters.post.PageRedirectionFilterTest```
```java
@Before
    public void setUp() {
        discoveryClient = mock(DiscoveryClient.class);

        RequestContext ctx = RequestContext.getCurrentContext();
        ctx.clear();
        ctx.set(SERVICE_ID_KEY, SERVICE_ID);
        MockHttpServletRequest request = new MockHttpServletRequest();
        response = new MockHttpServletResponse();
        ctx.setRequest(request);
        ctx.setResponse(response);

        GatewayConfigProperties gatewayConfigProperties = getGatewayConfigProperties(ctx);

        this.filter = new PageRedirectionFilter(this.discoveryClient, gatewayConfigProperties);
    }

//end of method
    private GatewayConfigProperties getGatewayConfigProperties(RequestContext ctx) {
        return GatewayConfigProperties.builder()
            .scheme(ctx.getRequest().getScheme())
            .hostname(ctx.getRequest().getLocalName() + ":" + ctx.getRequest().getLocalPort())
            .build();
    }
```

Reply:
Hi Ilkin,
Thanks for comments. I've modified code according to your comments in latest commit